WikiWatershed/model-my-watershed
# MapShed Raster Geoprocessing

## Overview

This PR creates a framework for performing MapShed related geoprocessing calls. The overall structure is as follows:

```
                                                                                  +--+
                                                                                     |
+---------------------------------------------------------------------------------+  |
| start_gwlfe_job                                                                 |  |
|                                                                                 |  |
| +-----------------------------------------------------------------------------+ |  |
| | geop_tasks                                                                  | |  |
| |                                                                             | |  |
| | +----------------+ +----------------+ +----------------+ +----------------+ | |  |
| | | mapshed_start  | | mapshed_start  | | mapshed_start  | |                | | |  |
| | |     -----      | |     -----      | |     -----      | |                | | |  |
| | | mapshed_finish | | mapshed_finish | | mapshed_finish | |      ...       | | |  |
| | |     -----      | |     -----      | |     -----      | |                | | |  |
| | |  nlcd_streams  | | post_process2  | | post_process3  | |                | | |  |
| | +----------------+ +----------------+ +----------------+ +----------------+ | |  |
| |                                                                             | |  |
| +-----------------------------------------------------------------------------+ |  |     +---------------------+
|                                                                                 |  |     |                     |
| +-----------------------------------------------------------------------------+ |  +---> |    save_job_error   |
| |                                                                             | |  |     |                     |
| |                                   identity                                  | |  |     +---------------------+
| |                                                                             | |  |
| +-----------------------------------------------------------------------------+ |  |
|                                                                                 |  |
| +-----------------------------------------------------------------------------+ |  |
| |                                                                             | |  |
| |                                 collect_data                                | |  |
| |                                                                             | |  |
| +-----------------------------------------------------------------------------+ |  |
|                                                                                 |  |
+---------------------------------------------------------------------------------+  |
                                                                                     |
+---------------------------------------------------------------------------------+  |
|                                                                                 |  |
|                                  save_job_result                                |  |
|                                                                                 |  |
+---------------------------------------------------------------------------------+  |
                                                                                     |
                                                                                  +--+
```

There is a group of geoprocessing chains, each of which is a triplet containing `mapshed_start`, `mapshed_finish`, and a post-processing task, which would be different for each call. The results of these are passed to an `identity` task, which simply acts as a workaround to https://github.com/celery/celery/issues/3191, and then fed into `collect_data`, which combines the geoprocessing results with values from the database. `save_job_result` saves the output, and `save_job_errors` handles any errors throughout the chain, just as in previous implementations.

MapShed related code has been refactored into its own module, since there are going to be a large number of Celery tasks related to each geoprocessing value, and it would be difficult to keep track of them in the already large `modeling.tasks`. Thus, the MapShed tasks have been moved from `modeling.tasks` to `modeling.mapshed.tasks`, and the calculation functions from `modeling.mapshed` to `modeling.mapshed.calcs`.

The structure of geoprocessing settings has been changed to accommodate multiple SJS endpoints and payloads.

We add a `streams` method which returns the streams contained in the polygon as GeoJSON. These streams are sent to the geoprocessing service so that we can get a list of NLCD codes for each stream pixel. From this, we calculate the ones categorized as agricultural, and count those as `AgLength`. The rest are counted as `UrbLength`.

See individual commits for details.

## Testing Instructions

Build the geoprocessing service from this PR https://github.com/WikiWatershed/mmw-geoprocessing/pull/32 and place in your worker VM. Restart the `spark-jobserver` service so it picks up the new JAR.

```bash
export MMWG=path/to/mmw-geoprocessing
export MMW=path/to/model-my-watershed

cd $MMWG/
./sbt "project summary" assembly
mv summary/target/scala-2.10/mmw-geoprocessing-assembly-1.0.0.jar $MMW/mmw-geoprocessing-1.0.0.jar
cd $MMW/
vagrant ssh worker -c 'mv -f /vagrant/mmw-geoprocessing-1.0.0.jar /opt/geoprocessing/ && sudo restart spark-jobserver'
```

Debug Celery to open an observable window:

```bash
./scripts/debugcelery.sh
```

In MMW [http://localhost:8000/](http://localhost:8000/), create a MapShed project and observe Celery. Everything should run successfully.

Connects #1258 
ASCII Art created in http://asciiflow.com/
